py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,424 @@
+    {end_time} The end time for the event.
+    {location} The location for the event.
+    {start_date} The start date for the event.
+    {start_time} The start time for the event.
+    {summary} The summary (i.e. title) for the event.
+    {time_to} The time until the event starts.
+
+format_time_to placeholders:
+    {days} The number of days until the event.
+    {hours} The number of hours until the event.
+    {minutes} The number of minutes until the event.
+
+Requires:
+    1. Python library google-api-python-client (can be obtained with pip).
+    2. OAuth 2.0 credentials for the Google Calendar api.
+
+    Follow Step 1 of the guide here to obtain your OAuth 2.0 credentials:
+    https://developers.google.com/google-apps/calendar/quickstart/python
+
+    Download the client_secret.json file which contains your client ID and
+    client secret. In your i3status config file, set configuration parameter
+    client_secret to the path to your client_secret.json file.
+
+    The first time you run the module, a browser window will open asking you
+    to authorize access to your calendar. After authorization is complete,
+    an access/refresh token will be saved to the path configured in
+    auth_token, and i3status will be restarted. This restart will
+    occur only once after the first time you successfully authorize.
+
+Example:
+```
+google_calendar {
+    client_secret = '/path/to/client_secret/'
+    auth_token = '/path/to/auth_token/'
+}
+```
+@author Igor Grebenkov
+@license BSD
+"""
+
+import httplib2
+import os
+import datetime
+
+from apiclient import discovery
+from oauth2client import client
+from oauth2client import tools
+from oauth2client.file import Storage
+from httplib2 import ServerNotFoundError
+
+
+SCOPES = 'https://www.googleapis.com/auth/calendar.readonly'
+APPLICATION_NAME = 'py3status google_calendar module'
+
+
+def gstr_to_datetime(date_time_str):
+    """ Returns a datetime object from a google calendar date/time string."""
+    tmp = '-'.join(date_time_str.split('-')[:-1])
+    return datetime.datetime.strptime(tmp, '%Y-%m-%dT%H:%M:%S')
+
+
+def datetime_to_str(date_time, dt_format):
+    """ Returns a strftime formatted string from a datetime object."""
+    return date_time.strftime(dt_format)
+
+
+def delta_time(date_time):
+    """
+    Returns in a dict the number of days/hours/minutes and total minutes
+    until date_time.
+    """
+    now = datetime.datetime.now()
+    diff = date_time - now
+
+    days = int(diff.days)
+    hours = int(diff.seconds / 3600)
+    minutes = int((diff.seconds / 60) - (hours * 60)) + 1
+    total_minutes = int((diff.seconds / 60) + (days * 24 * 60)) + 1
+
+    return {'days': days,
+            'hours': hours,
+            'minutes': minutes,
+            'total_minutes': total_minutes}
+
+
+class Py3status:
+    auth_token = '~/.config/py3status/google_calendar.json'
+    button_open = 3
+    button_toggle = 1
+    cache_timeout = 60
+    client_secret = ''
```

Comment:
From what I gather, we need two things to use this... `auth_token` and `client_secret`

We default `auth_token` and not `client_secret`. Maybe we do this too?
* `~/.config/py3status/google_calendar.auth_token`
* `~/.config/py3status/google_calendar.client_secret`

Reply:
Good idea. I've made these the defaults. No more hardcoded `client_secret.json`.

So you couldn't get it to fetch events? Is that error when you run the module from the terminal or in the py3status logs?

Reply:
I don't have problems fetching events if I used the [quickstart.py](https://developers.google.com/google-apps/calendar/quickstart/python). Running `google_calendar` in terminal (Python2 and Python3) and running this in py3status bar all will give me the same error. I'm pinging @tobes in case it's a composite issue not related to the modules. The error is this.
```
1d [chris:~/src/py3status/py3status/modules] igorgrebenkov-addmodule_google_calendar(2)+ ± python3 google_calendar.py
'Composite' object has no attribute 'strip' (AttributeError) composite.py line 99.
Traceback
AttributeError: 'Composite' object has no attribute 'strip'
  File "/home/chris/src/py3status/py3status/module.py", line 751, in run
    self.process_composite(response)
  File "/home/chris/src/py3status/py3status/module.py", line 350, in process_composite
    composite = composite.simplify().get_content()
  File "/home/chris/src/py3status/py3status/composite.py", line 99, in simplify
    if diff == diff_last or (item['full_text'].strip() == '' and item_last):
```

Reply:
Hmm, okay. I'm reworking how I build the composite anyways at the moment, since the current approach isn't working with color thresholds. It seems that when `\?color=threshold' is in the event format string, when it's used with self.py3.format, it creates a composite?

Reply:
```
364           responses.append({
365               'full_text': event_formatted,
366               'color': self.colors[index], 'index': index})

```
I think it should be `responses.append(event_formatted)` since you safeformatted it earlier... and maybe not passing in color/index. There are also an issue of doing separators right. I'm familiar with making composites with `safe_format` and not so much with `composite_create`.

Reply:
Yes, I was going to tackle separators after this. Before, `event_formatted` was just a string after the `safe_format` so the above worked (at least for me, although the approach kinda sucks anyway). With color thresholds, `event_formatted` is now a composite after `safe_format`, I guess since it needs to apply the color. So when I add it to the list of responses like you mention above, `responses.append(event_formatted)`, I end up with `responses` as a list of composites, which gives me invalid format when I try to make a composite out of that and display it since then I have a composite of composites. 

I tried `composite_join` thinking it would give me one composite that's the union of two composites, but it doesn't seem to work that way.

Reply:
Hard to review when I can't use it here. Lulz. I think replacing...
* `return {'events': self.py3.composite_create(responses)}` with...
* `return {'events': responses}` if you're going to use `safe_format`.

Reply:
Ohh, I think I get it now! `composite_join` will take a list of composites and join them into one composite with the input separator. So it solves the issue I was having and separators. Great!